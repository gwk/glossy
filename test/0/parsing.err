note: `main`: minimized DFA contains 5 post-match nodes.
