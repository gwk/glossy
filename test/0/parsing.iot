{
  'args': [
    '-test',
    '10 0b10 0q10 0o10 0d10 0x10',
    '0x__ffff_ffff_ffff_ffff',
    #'0x1_0000_0000_0000_0000', # temporarily disabled because swift overflows but python does not.
  ],
  'code': 0,
}
