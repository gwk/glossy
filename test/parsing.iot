{
  'skip': True,
  'compile': ['$PROJ/legs/__main__.py $STEM.legs -test -output $NAME.swift'],
  'cmd': './$NAME.swift',
  'args': [
    'a (bc def) g',
    '/* comment "contents" () / *. */',
    "'single quoted.' \"double quoted.\"",
    '/* nested /* comment. */ */',
    "'sq \(a) interpolated.'",
    '"dq \(a) interpolated."',
    "'sq \(nested \"interpolated \(a)\").'",
  ],
}