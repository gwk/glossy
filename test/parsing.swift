// Dedicated to the public domain under CC0: https://creativecommons.org/publicdomain/zero/1.0/.

import Foundation


func main() {
  for (i, arg) in CommandLine.arguments.enumerated() {
    if i == 0 { continue }
    test(index: i, arg: arg)
  }
}


func test(index: Int, arg: String) {
  let name = "arg\(index)"
  print("\n\(name): \(arg.repr)")
  let data = Data(arg.utf8)
  let source = Source(name: name, data: data)
  for token in source.lex() {
    let base: Int
    var from = 2 // "0_" prefix is the common case.
    switch token.kind {
      case .space: continue
      case .incomplete, .invalid: base = 0
      case .num:  base = 10; from = 0
      case .bin:  base = 2
      case .quat: base = 4
      case .oct:  base = 8
      case .dec:  base = 10
      case .hex:  base = 16
    }
    var msg: String
    do {
      let val = try source.parseDigits(token: token, from: from, base: base)
      msg = "\(token.kind): \(val)"
    } catch let e {
      msg = "error: \(e)"
    }
    let d = source.diagnostic(token: token, prefix: "token", msg: msg, showMissingFinalNewline: false)
    print(d, terminator: "")
  }
}


extension String {
  var repr: String {
    var r = "\""
    for char in unicodeScalars {
      switch char {
      case "\\": r.append("\\\\")
      case "\"": r.append("\\\"")
      case UnicodeScalar(0x20)...UnicodeScalar(0x7E): r.append(String(char))
      case "\0": r.append("\\0")
      case "\t": r.append("\\t")
      case "\n": r.append("\\n")
      case "\r": r.append("\\r")
      default: r.append("\\{\(String(char.value, radix: 16, uppercase: false))}")
      }
    }
    r.append("\"")
    return r
  }
}
