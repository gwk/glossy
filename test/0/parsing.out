
arg1: '10 0b10 0q10 0o10 0d10 0x10'
arg1:1:1-3: num: 10
10 0b10 0q10 0o10 0d10 0x10
~~
arg1:1:3-4: space
10 0b10 0q10 0o10 0d10 0x10
  ~
arg1:1:4-8: bin: 2
10 0b10 0q10 0o10 0d10 0x10
   ~~~~
arg1:1:8-9: space
10 0b10 0q10 0o10 0d10 0x10
       ~
arg1:1:9-13: quat: 4
10 0b10 0q10 0o10 0d10 0x10
        ~~~~
arg1:1:13-14: space
10 0b10 0q10 0o10 0d10 0x10
            ~
arg1:1:14-18: oct: 8
10 0b10 0q10 0o10 0d10 0x10
             ~~~~
arg1:1:18-19: space
10 0b10 0q10 0o10 0d10 0x10
                 ~
arg1:1:19-23: dec: 10
10 0b10 0q10 0o10 0d10 0x10
                  ~~~~
arg1:1:23-24: space
10 0b10 0q10 0o10 0d10 0x10
                      ~
arg1:1:24-28: hex: 16
10 0b10 0q10 0o10 0d10 0x10
                       ~~~~

arg2: '0x__ffff_ffff_ffff_ffff'
arg2:1:1-24: hex: 18446744073709551615
0x__ffff_ffff_ffff_ffff
~~~~~~~~~~~~~~~~~~~~~~~

arg3: '0x1_0000_0000_0000_0000'
arg3:1:1-24: error: overflow(hex:0-23)
0x1_0000_0000_0000_0000
~~~~~~~~~~~~~~~~~~~~~~~
